MCPClient: single-threaded implementation

sevein opened the pull request:
[RDSSARK-374](https://jiscdev.atlassian.net/browse/RDSSARK-374)

This pull request introduces a simplified version of MCPClient where threaded workers are replaced by a single worker loop in the main thread. This loop is now forced to exit by a signal handler `quit` when `SIGINT` or `SIGTERM` signals are received.

This implementation forces the worker to exit immediately which is potentially better than having ECS killing it after the 30 seconds timeout is exceeded. This causes the running jobs to stop and Gearman will eventually schedule them to the next available worker.

jrwdunham committed:
MCPClient: refactor archivematicaClient.py

Adds docstrings, more Pythonic formatting, and some cosmetic structural
changes to archivematicaClient.py

sevein committed:
MCPClient: use exec form in Docker entrypoint

This commit updates the `ENTRYPOINT` so it uses the `exec` form instead of the
`shell` form. This avoids having an intermediate shell between the Docker
daemon and the Python interpreter which allows us to receive all the system
signals.

lower29 commented on src/MCPClient.Dockerfile:
```diff
@@ -99,4 +99,4 @@ RUN mkdir -p /etc/archivematica && (echo "---"; \
 USER archivematica
 
-ENTRYPOINT /src/MCPClient/lib/archivematicaClient.py
+ENTRYPOINT ["/src/MCPClient/lib/archivematicaClient.py"]
```
This is neat. The difference in meaning is subtle; there are probably other Docker images in our stack that could benefit from a change like this too.

lower29 commented on src/MCPClient/lib/archivematicaClient.py:
```diff
@@ -21,173 +47,197 @@
+        arguments = arguments.replace(var, val)
+    arguments = arguments.replace('%taskUUID%', task_uuid)
+    script = client_script_full_path + ' ' + arguments
+    return script, task_uuid
```
The docstring says that this function returns a 3-tuple of values, but here we are only returning the `script` and `task_uuid`. What happened to the capture output boolean, or was that aspirational? I think we should either update the docstring or add the missing boolean to the return value (and use it in `execute_command` in the call to `executeOrRun`), either way it needs changing.

sevein replied:
Good catch @lower29, I'll fix!

sevein replied:
> What happened to the capture output boolean, or was that aspirational?

It's something that Joel has been adding to the vanilla 1.8 WIP branch. I guess he forgot to delete it when he ported the refactor to 1.7.

lower29 reviewed (changes requested):
I really like the new code from @jrwdunham, a great job on the refactor - much cleaner and easier to understand.

The transition from a threaded app to single-threaded looks good too - is this something we will want for all deployments of MCPClient going forward or just something for Jisc? I wonder if maybe we should be retaining the threaded version for other deployment scenarios? For example when there isn't a watchdog to ensure the mcpclient service(s) is still running.

I spotted a minor thing in the `_process_gearman_job` function that I think needs fixing - an inconsistency between the docs and the actual implementation.

My other concern is that we're removing a lib file from `archivematicaCommon`, because it's no longer used in `archivematicaClient.py`. Is this only place where the `detectCores` was being used? If so then fine, we can resurrect from history if it's needed again, otherwise it should be left in.

sevein commented:
> The transition from a threaded app to single-threaded looks good too - is this something we will want for all deployments of MCPClient going forward or just something for Jisc? I wonder if maybe we should be retaining the threaded version for other deployment scenarios? For example when there isn't a watchdog to ensure the mcpclient service(s) is still running.

That's actually a very good idea. If I add the single-threaded version as a new executable, e.g. `archivematicaSingleClient.py`, then we'll be able to upstream this work to vanilla in a backward-compatible manner. I'll do that, it shouldn't be hard. Thanks for the idea.

> My other concern is that we're removing a lib file from archivematicaCommon, because it's no longer used in archivematicaClient.py. Is this only place where the detectCores was being used? If so then fine, we can resurrect from history if it's needed again, otherwise it should be left in.

I don't delete code unless I confirm it's dead code but with your suggestion if the old entry point remains I'll also have to restore that module.

Thank you so much!

sevein commented:
I felt very tempted to write this in Go by the way. It's funny that I've spent all this time writing web applications in Python but when it comes to concurrency topics I find Go much easier because of its channels or things like `context` or `select`.

I guess it's never too late to learn more about multi-threaded programming and the synchronization mechanisms provided by the `threading` module in Python - I'm a total illiterate at that! One of my first attempts was to update the multi-threaded version so it stops accepting new jobs as long as the signal is received but also wait up to 30 secs for the ongoing jobs to complete before killing them. I also wanted to signal the subprocesses so they could also try to exit gracefully. I tried a few things but I failed badly. Multi-threaded programming is extremely difficult!

sevein committed:
MCPClient: add single-threaded client

This commit adds a new executable `archivematicaSingleClient.py` that can be
used to execute a single-threaded MCPClient worker. This commit also updates
the Dockerfile so the new ENTRYPOINT points to this executable, this is because
Docker container are frequently executed by a scheduler where it is easy to
manage multiple replicas.

sevein commented:
@lower29 I've just submitted a new commit where I add a new entry point instead of making the changes in the original one. The diff is now bigger because I moved things around a bit to avoid duplication of code. It's probably easier if you look at the tree directly instead of deciphering the diff.

sevein commented on src/MCPClient.Dockerfile:
```diff
@@ -99,4 +99,4 @@ RUN mkdir -p /etc/archivematica && (echo "---"; \
 USER archivematica
 
-ENTRYPOINT /src/MCPClient/lib/archivematicaClient.py
+ENTRYPOINT ["/src/MCPClient/lib/archivematicaSingleClient.py"]
```
Two things are happening here:
- Use the `exec` form instead of the `shell` form so we can capture signals from the application.
- Use the new `archivematicaSingleClient.py` executable which is the new single-threaded implementation.

lower29 commented:
I've re-reviewed @sevein's changes and have run a test transfer through the new MCP client code to confirm it works. Happy to approve 👍
